src/REPLMode.jl

```diff
@@ -267,76 +274,139 @@ function _statement(words)
         (:arg, word, command, !manifest)
 end
 
-function parse(cmd::String; for_completions=false)
-    # replace new lines with ; to support multiline commands
-    cmd = replace(replace(cmd, "\r\n" => "; "), "\n" => "; ")
-    # tokenize accoring to whitespace / quotes
-    qwords = parse_quotes(cmd)
-    # tokenzie unquoted tokens according to pkg REPL syntax
-    words = lex(qwords)
-    # break up words according to ";"(doing this early makes subsequent processing easier)
-    word_groups = group_words(words)
-    # create statements
-    if for_completions
-        return _statement(word_groups[end])
+tokenize(command::AbstractString) =
+    lex(parse_quotes(command))
+
+#TODO this could be a utility?
+function chunk(tokens::Vector{String})
+    chunks = Vector{String}[]
+    chunk = String[]
+    for token in tokens
+        if token == ";"
+            # TODO should error on empty command?
+            if !isempty(chunk)
+                push!(chunks, chunk)
+                chunk = String[]
+            end
+        else
+            push!(chunk, token)
+        end
     end
-    return map(Statement, word_groups)
+    !isempty(chunk) && push!(chunks, chunk)
+
+    isempty(chunks) && repl_error(ERROR_NO_INPUT)
+    return chunks
 end
 
-# vector of words -> structured statement
-# minimal checking is done in this phase
-function Statement(words)::Statement
-    is_option(word) = first(word) == '-'
-    statement = Statement()
+preprocess(input::String) =
+    replace(replace(input, "\r\n" => "; "), "\n" => "; ")
 
-    word = popfirst!(words)
-    # meta options
-    while is_option(word)
-        push!(statement.meta_options, word)
-        isempty(words) && pkgerror("no command specified")
-        word = popfirst!(words)
+function parse(input::String)
+    verbose = nothing
+    try
+        tokens = tokenize(preprocess(input))
+        commands = chunk(tokens)
+        statements = Statement[]
+        for command in commands
+            verbose = "(when parsing `$(join(command, " "))`)"
+            push!(statements, Statement(command))
+        end
+        return statements
+    catch ex
+        (ex isa PkgError && ex.class == PKG_ERROR_REPL) || rethrow()
```

Comment:
I'm curious why you suggest this? Is there an advantage? I actually had `REPLError` in there before but I switched to using `PkgError` for everything. In any case, I don't have strong feelings about either option so I can just switch it back.

Reply:
It doesn't really matter but it seems odd to much with the internal fields of the PkgError here instead of having the error itself take care of that. Since this is specific for the REPL is why I suggested having a dedicated REPLError type that can evolve with the REPL and is "shielded" from the rest of the code base.